// matter-lines/lib/js/matter-line.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _line = require("./line");

var _line2 = _interopRequireDefault(_line);

var _matterJs = require("matter-js");

var _events = require("events");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var MatterLine = function (_Line) {
  _inherits(MatterLine, _Line);

  function MatterLine(world) {
    var defaultPoints = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
    var lineWidth = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 10;
    var miterLimit = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 10;

    _classCallCheck(this, MatterLine);

    var _this = _possibleConstructorReturn(this, (MatterLine.__proto__ || Object.getPrototypeOf(MatterLine)).call(this, defaultPoints, lineWidth, miterLimit));

    _this.world = world;
    _this.body = null;
    _this.generateBody();
    return _this;
  }

  _createClass(MatterLine, [{
    key: "addPoint",
    value: function addPoint(point) {
      _get(MatterLine.prototype.__proto__ || Object.getPrototypeOf(MatterLine.prototype), "addPoint", this).call(this, point);
      this.generateBody();
    }
  }, {
    key: "generateBody",
    value: function generateBody() {
      var vertices = this.getVertices();
      if (vertices.length <= 2) {
        if (this.body !== null) {
          _matterJs.World.remove(this.world, this.body);
        }
        this.body = null;
      } else {
        if (this.body !== null) {
          _matterJs.World.remove(this.world, this.body);
        }
        this.body = _matterJs.Bodies.fromVertices(0, 0, vertices);
        _matterJs.World.add(this.world, this.body);
        var bounds = _matterJs.Bounds.create(vertices);
        _matterJs.Body.setPosition(this.body, {
          x: this.body.position.x - this.body.bounds.min.x + bounds.min.x,
          y: this.body.position.y - this.body.bounds.min.y + bounds.min.y
        });
        _matterJs.Body.setStatic(this.body, true);
      }
    }
  }]);

  return MatterLine;
}(_line2.default);

exports.default = MatterLine;
